### PR TITLE
Issue #1265: AddODataQueryFilter

### PR DESCRIPTION
**Pull request by biaol-odata**
### Issues

*This pull request adds test coverage to the correct processing of AddODataQueryFilter.*

### Description

*For classic and netcore, adding tests to confirm that AddODataQueryFilter works in lieu of EnableQuery Attribute on controller.*

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/ODataContainmentTest.cs [hunk 1]
```diff
@@ -503,6 +528,15 @@ public SingleResult<MyOrder> Get(int orderId)
                 }
             }
 
+//            [EnableQuery]      // Using _configuration.AddODataQueryFilter
```
Comment (AlanWong-MS) [inline test/UnitTest/Microsoft.AspNet.OData.Test.Shared/ODataContainmentTest.cs]: indentation

File: test/UnitTest/Microsoft.AspNet.OData.Test.Shared/ODataContainmentTest.cs [hunk 2]
```diff
@@ -503,6 +528,15 @@ public SingleResult<MyOrder> Get(int orderId)
...
+            [ODataRoute("MyWackyOrders({orderId})")]
+            public SingleResult<MyOrder> GetViaProgramaticEnableQuery(int orderId)
+            {
+                // Wacky Get query by (orderId+1)
+                var result = _myOrders.AsQueryable().Where(mo => mo.ID == (orderId+1));
```
Comment (AlanWong-MS) [inline test/UnitTest/Microsoft.AspNet.OData.Test.Shared/ODataContainmentTest.cs]: nit: spacing in "orderId + 1"

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Formatter/JsonLight/Metadata/ComplexTypeTests.cs [hunk 3]
```diff
@@ -85,5 +86,38 @@ protected static IEdmModel GetEdmModel(WebRouteConfiguration configuration)
                 }
             }
         }
+
+        [Theory]
+        [InlineData("application/json;odata.metadata=full", "/Complex/EntityWithComplexProperties?$select=StringListProperty")]
+        [InlineData("application/json;odata.metadata=full", "/Complex/EntityWithComplexProperties?$select=ComplexProperty")]
+        [InlineData("application/json;odata.metadata=full", "/Complex/EntityWithComplexProperties?$count=true")]
+        [InlineData("application/json;odata.metadata=full", "/Complex/EntityWithComplexProperties?$count=true&$select=ComplexProperty")]
```
Comment (AlanWong-MS) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Formatter/JsonLight/Metadata/ComplexTypeTests.cs]: The title of your PR says AddODataQueryFilter. Is the $filter implementation not included for this test case?

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Formatter/JsonLight/Metadata/ComplexTypeTests.cs [hunk 4]
```diff
@@ -85,5 +86,38 @@ protected static IEdmModel GetEdmModel(WebRouteConfiguration configuration)
...
+
+        /* Error: "A path within the select or expand query option is not supported."
+        [InlineData("application/json;odata.metadata=full", "/Complex/EntityWithComplexProperties?$select=ComplexProperty/StringListProperty")]
+        */
+
+        /* Error: "Property 'ComplexProperty' on type 'Microsoft.Test.E2E.AspNet.OData.Formatter.JsonLight.Metadata.Model.EntityWithComplexProperties' is not a navigation property or complex property. Only navigation properties can be expanded."
+        [InlineData("application/json;odata.metadata=full", "/Complex/EntityWithComplexProperties?$expand=ComplexProperty")]
+        */
+
+        public async Task ODataTypeAnnotationShouldAppearForComplexTypesLikeCollectionAndUserDefinedTypes_xyz(string acceptHeader, string uriPath)
```
Comment (AlanWong-MS) [inline test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Formatter/JsonLight/Metadata/ComplexTypeTests.cs]: I think you might have accidentally left the "_xyz" in the test case name.

**Comment by biaol-odata** [comment]
I think I should close this PR for now. Test code only.

File: test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Formatter/JsonLight/Metadata/ComplexTypeTests.cs (hunk 3 above)
Comment (biaol-odata) [reply test/E2ETest/Microsoft.Test.E2E.AspNet.OData/Formatter/JsonLight/Metadata/ComplexTypeTests.cs]: The AddODataQueryFilter is for enable the service configuration for post-execution query processing. It is not really for $filter.

**Comment by ramil-mux** [comment]
Hello, @biaol-odata!
Could you please explain if the EnableQuery attribute for the GET controller method suppresses the configurations in method HttpConfigurationExtensions.AddODataQueryFilter in the classic asp.net web application?
If I set property MaxNodeCount = 1000 in AddODataQueryFilter and use EnableQuery attribute in controller method this setting doesn't work, but if I delete EnableQuery attribute property work.
**WebApiConfig:**
```
            config.AddODataQueryFilter(new EnableQueryAttribute()
            {
                PageSize = 500,
                MaxNodeCount = 1000,
                MaxOrderByNodeCount = 15
            });
```
**Controller:**
```
        [EnableQuery]
        public IQueryable<T> Get()
        {
            var ret = _entityService.GetEntities().AsNoTracking();
            return ret;
        }
```

**Pull request closed without merging**